sourced/ml/uast_struct_to_bag.py

```diff
@@ -0,0 +1,164 @@
+from collections import defaultdict
+from random import random, choice
+
+from sourced.ml.uast_ids_to_bag import FakeVocabulary
+
+
+class UASTStructure2BagBase:
+    def uast_to_bag(self, uast):
+        raise NotImplemented
+
+
+class UAST2Seq2Bag(UASTStructure2BagBase):
+    """
+    DFS traversal + preserve order of node children.
+    """
+    def __init__(self, type2ind=None, stride=1, seq_len=5):
+        self.type2ind = type2ind if type2ind is not None else FakeVocabulary()
+        self.stride = stride
+        self.seq_len = seq_len
+
+    def _uast2seq(self, root, walk):
+        walk.append(root)
+        for ch in root.children:
+            self._uast2seq(ch, walk)
+
+    def uast_to_bag(self, uast):
+        bag = defaultdict(int)
+        seq = []
+        self._uast2seq(uast, seq)
+
+        seq = [self.node2ind(n) for n in seq]
+
+        for i in range(0, len(seq) - self.seq_len, self.stride):
+            # convert to str - requirement from wmhash.BagsExtractor
+            bag[str(seq[i:i + self.seq_len])] += 1
+        return bag
```

Comment:
Please place all the keys into `w.` namespace (see `IdentifiersBagExtractor`)

Reply:
Right now it's a part of `BagsExtractor` functionality - so every extractor should be provided with `namespace` argument during initialization.